content/filterx/_index.md

```diff
@@ -0,0 +1,379 @@
+- Pipeline variables. These are simliar to local variables, but must be declared before first use, for example, `declare my-pipeline-variable=5;`
+
+    Pipeline variables are available in the current and all subsequent filterx block. They're global in the sense that you can access them from multiple filterx blocks, but note that they're still attached to the particular message that is processed, it's value isn't preserved between messages.
+
+{{% alert title="Note" color="info" %}}
+If you want to pass data between two filterx blocks of a log statement, use pipeline variables. That has better performance than name-value pairs.
+{{% /alert %}}
+
+## Variable types
+
+Variables can have the following types:
+
+- `boolean`
+- `bytes` (to represent binary data)
+- `datetime`
+- `dict`
+- `double`
+- `int`
+- `json_object` and `json_array` for JSON or JSON-like objects. The `json` type is an alias for the `json_object` type. <!-- object: {}, array []-->
+- `list`
+- `protobuf`
+- `string`
+<!-- - otel... types -->
+
+### Assign values
+
+To assign value to a name-value pair or a variable, use the following syntax:
+
+```shell
+<variable-name> = <value-of-the-variable>;
+```
+
+Usually you can omit the type, and {{< product >}} automatically assigns the type based on the syntax of the value, for example:
+
+- `mystring = "string-value";`
+- `myint = 3;`
+- `mydouble = 2.5;`
+- `myboolean = true;`
+
+If needed, you can explicitly specify the type of the variable:
+
+```shell
+<variable-name> = <variable-type>(<value-of-the-variable>);
+```
+
+For example:
+
+```shell
+filterx {
+  ${MESSAGE} = string("Example string message");
+};
+```
+
+You can also assign the value of other name-value pairs, for example:
+
+```shell
+filterx {
+  ${MESSAGE} = ${HOST};
+};
+```
+
+{{% alert title="Note" color="info" %}}
+When assigning values to name-value pairs, you cannot modify [hard macros]({{< relref "/chapter-manipulating-messages/customizing-message-format/macros-hard-vs-soft/_index.md" >}}).
+{{% /alert %}}
+
+You can use the traditional [template functions]({{< relref "/chapter-manipulating-messages/customizing-message-format/reference-template-functions/_index.md" >}}) of {{< product >}} to access and format name-value pairs. Note that you must enclose the template function expression between double-quotes, for example:
+
+```shell
+${MESSAGE} = "$(format-json --subkeys values.)";
+```
+
+### Delete values
+
+To delete a value without deleting the object (for example, name-value pair), use the `null value`, for example:
+
+```shell
+${MY-NV-PAIR} = null;
+```
+
+To delete the name-value pair (or a key from an object), use the `unset` function:
+
+```shell
+unset(${MY-NV-PAIR});
```

Comment:
Is it possible to unset builtins like `MESSAGE`?

Reply:
I asked for clarification from @alltilla , and the rule is hard macros cannot be unset. And nor be set.

We could reference the hard vs soft macros here: https://axoflow.com/docs/axosyslog-core/chapter-manipulating-messages/customizing-message-format/macros-hard-vs-soft/_print/?highlight=hard&highlight=macro

And add a warning/info about the assignment limitation too.